[GTK] Printing not working with bubblewrap sandbox enabled

carlosgcampos opened the pull request:
#### d5158b635f0e4ec3842ed6b0ae259d2f33b00dd4
<pre>
[GTK] Printing not working with bubblewrap sandbox enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=202363">https://bugs.webkit.org/show_bug.cgi?id=202363</a>

Reviewed by Michael Catanzaro.

Move printing implementation from the web process to the UI process. The
web process now writes the resulting document to a buffer that is sent
as shared memory to the UI process. The UI process sends the document to
the printer using GtkPrintJob API.

* Source/WebKit/Shared/PrintInfo.h: The constructor now receives a GtkPrintJob.
* Source/WebKit/Shared/WebErrors.h: Remove printing errors that can&apos;t
happen in the web process now.
* Source/WebKit/Shared/gtk/ArgumentCodersGtk.cpp:
(IPC::encode): Handle null print settings and page setup.
(IPC::decode): Ditto.
* Source/WebKit/Shared/gtk/PrintInfoGtk.cpp:
(WebKit::PrintInfo::PrintInfo): Create new settings populated from the GtkPrintJob.
* Source/WebKit/Shared/gtk/WebErrorsGtk.cpp:
(WebKit::printError): Deleted.
(WebKit::printerNotFoundError): Deleted.
* Source/WebKit/SourcesGTK.txt: Remove PrinterListGtk.
* Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
(webkitPrintOperationRunDialog): Save the selected printer too.
(webkitPrintOperationFailed): Helpers to emit the failed signals for the given errors.
(webkitPrintOperationFinished): Helper to emit the finished signal and clean up the operation.
(webkitPrintOperationPrintPagesForFrame): Create a GtkPrintJob for the operation and wait until
the web process has written the document to send it to the printer.
(webkitPrintOperationRunDialogForFrame): Pass the printer to webkitPrintOperationPrintPagesForFrame() too.
(printerFromSettingsOrDefault): Helper to get the GtkPrinter for the given settings or the default printer if any.
(webkit_print_operation_print): Pass the printer to webkitPrintOperationPrintPagesForFrame() too.
(drawPagesForPrintingCompleted): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::drawPagesForPrinting): Remove the callback wrapper.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::print): Remove the printer list hack.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::close): disconnectFromPage() is no longer needed.
(WebKit::WebPage::beginPrinting): Use makeUnique to create the WebPrintOperationGtk.
(WebKit::WebPage::drawPagesForPrinting): Copy the shared buffer to shared memory and send the handle to the UI process.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/gtk/PrinterListGtk.cpp: Removed.
(WebKit::PrinterListGtk::getOrCreate): Deleted.
(WebKit::PrinterListGtk::PrinterListGtk): Deleted.
(WebKit::PrinterListGtk::~PrinterListGtk): Deleted.
(WebKit::PrinterListGtk::addPrinter): Deleted.
(WebKit::PrinterListGtk::findPrinter const): Deleted.
* Source/WebKit/WebProcess/WebPage/gtk/PrinterListGtk.h: Removed.
* Source/WebKit/WebProcess/WebPage/gtk/WebPrintOperationGtk.cpp:
(WebKit::WebPrintOperationGtk::PrintPagesData::PrintPagesData): Make PrintPagesData a private struct in WebPrintOperationGtk
to simplify everyting.
(WebKit::WebPrintOperationGtk::PrintPagesData::incrementPageSequence):
(WebKit::WebPrintOperationGtk::WebPrintOperationGtk):
(WebKit::WebPrintOperationGtk::~WebPrintOperationGtk):
(WebKit::WebPrintOperationGtk::startPrint): Create the cairo surface according to the settings to be written on a stream.
(WebKit::WebPrintOperationGtk::startPage):
(WebKit::WebPrintOperationGtk::endPage):
(WebKit::WebPrintOperationGtk::endPrint):
(WebKit::WebPrintOperationGtk::printPagesIdle):
(WebKit::WebPrintOperationGtk::printPagesIdleDone):
(WebKit::WebPrintOperationGtk::printPagesDone):
(WebKit::WebPrintOperationGtk::printDone):
(WebKit::WebPrintOperationGtk::print):
(WebKit::PrintPagesData::PrintPagesData): Deleted.
(WebKit::PrintPagesData::collatedCopiesLeft): Deleted.
(WebKit::PrintPagesData::uncollatedCopiesLeft): Deleted.
(WebKit::PrintPagesData::copiesLeft): Deleted.
(WebKit::PrintPagesData::incrementPageSequence): Deleted.
(WebKit::WebPrintOperationGtk::create): Deleted.
(WebKit::WebPrintOperationGtk::disconnectFromPage): Deleted.
* Source/WebKit/WebProcess/WebPage/gtk/WebPrintOperationGtk.h:
(WebKit::WebPrintOperationGtk::PrintPagesData::collatedCopiesLeft const):
(WebKit::WebPrintOperationGtk::PrintPagesData::uncollatedCopiesLeft const):
(WebKit::WebPrintOperationGtk::PrintPagesData::copiesLeft const):
(WebKit::WebPrintOperationGtk::printContext const): Deleted.
(WebKit::WebPrintOperationGtk::printSettings const): Deleted.
(WebKit::WebPrintOperationGtk::pageSetup const): Deleted.
(WebKit::WebPrintOperationGtk::printMode const): Deleted.
(WebKit::WebPrintOperationGtk::setNumberOfPagesToPrint): Deleted.
(WebKit::WebPrintOperationGtk::pagesToPrint const): Deleted.
(WebKit::WebPrintOperationGtk::pagePosition const): Deleted.
(WebKit::WebPrintOperationGtk::setPagePosition): Deleted.
(WebKit::WebPrintOperationGtk::pageRanges const): Deleted.
(WebKit::WebPrintOperationGtk::pageRangesCount const): Deleted.
(WebKit::WebPrintOperationGtk::numberUp const): Deleted.
(WebKit::WebPrintOperationGtk::numberUpLayout const): Deleted.
(WebKit::WebPrintOperationGtk::pageSet const): Deleted.
(WebKit::WebPrintOperationGtk::reverse const): Deleted.
(WebKit::WebPrintOperationGtk::copies const): Deleted.
(WebKit::WebPrintOperationGtk::collateCopies const): Deleted.
(WebKit::WebPrintOperationGtk::scale const): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitGtk/TestPrinting.cpp:
(findPrintToFilePrinter):
(testPrintOperationPrint):
(testPrintOperationErrors):
(testPrintOperationCloseAfterPrint):
(testPrintCustomWidget):
(testPrintOperationPrintPrinter): Deleted.

Canonical link: <a href="https://commits.webkit.org/253895@main">https://commits.webkit.org/253895@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9ca66c42d61761314685064d94025f83603db784

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87172 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31259 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18003 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96186 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149745 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91148 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29623 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/25862 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79292 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91185 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92788 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/23926 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/73979 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23832 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/78908 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79155 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66847 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27348 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/12991 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27295 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14005 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2737 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28980 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/36863 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28919 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33282 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -309,29 +317,78 @@ static WebKitPrintOperationResponse webkitPrintOperationRunDialog(WebKitPrintOpe
 #endif
 
-static void drawPagesForPrintingCompleted(API::Error* wkPrintError, WebKitPrintOperation* printOperation)
+static void webkitPrintOperationFailed(WebKitPrintOperation* printOperation, GError* error)
```
I see it's convenient to be able to pass the result of g_error_new_literal() directly to this function, but passing ownership of `error` through a raw pointer is still not cool. I guess I would take a `GRefPtr<GError>&&` here, then you don't need much adjustment to the callsites and can drop the manual g_error_free().

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -309,29 +317,78 @@ static WebKitPrintOperationResponse webkitPrintOperationRunDialog(WebKitPrintOpe
-        drawPagesForPrintingCompleted(printError, adoptGRef(printOperation).get());
+    page.drawPagesForPrinting(webFrame, printInfo, [printOperation = GRefPtr<WebKitPrintOperation>(printOperation)](std::optional<SharedMemory::IPCHandle>&& data, WebCore::ResourceError&& error) mutable {
+        auto* priv = printOperation->priv;
+        // When running synchronously WebPageProxy::printFrame() calls endPrinting().
```
```suggestion
        // When running synchronously, WebPageProxy::printFrame() calls endPrinting().
```

mcatanzaro reviewed (approved):
This is great, thanks.

carlosgcampos committed:
[GTK] Printing not working with bubblewrap sandbox enabled
https://bugs.webkit.org/show_bug.cgi?id=202363

Reviewed by Michael Catanzaro.

Move printing implementation from the web process to the UI process. The
web process now writes the resulting document to a buffer that is sent
as shared memory to the UI process. The UI process sends the document to
the printer using GtkPrintJob API.

* Source/WebKit/Shared/PrintInfo.h: The constructor now receives a GtkPrintJob.
* Source/WebKit/Shared/WebErrors.h: Remove printing errors that can't
happen in the web process now.
* Source/WebKit/Shared/gtk/ArgumentCodersGtk.cpp:
(IPC::encode): Handle null print settings and page setup.
(IPC::decode): Ditto.
* Source/WebKit/Shared/gtk/PrintInfoGtk.cpp:
(WebKit::PrintInfo::PrintInfo): Create new settings populated from the GtkPrintJob.
* Source/WebKit/Shared/gtk/WebErrorsGtk.cpp:
(WebKit::printError): Deleted.
(WebKit::printerNotFoundError): Deleted.
* Source/WebKit/SourcesGTK.txt: Remove PrinterListGtk.
* Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
(webkitPrintOperationRunDialog): Save the selected printer too.
(webkitPrintOperationFailed): Helpers to emit the failed signals for the given errors.
(webkitPrintOperationFinished): Helper to emit the finished signal and clean up the operation.
(webkitPrintOperationPrintPagesForFrame): Create a GtkPrintJob for the operation and wait until
the web process has written the document to send it to the printer.
(webkitPrintOperationRunDialogForFrame): Pass the printer to webkitPrintOperationPrintPagesForFrame() too.
(printerFromSettingsOrDefault): Helper to get the GtkPrinter for the given settings or the default printer if any.
(webkit_print_operation_print): Pass the printer to webkitPrintOperationPrintPagesForFrame() too.
(drawPagesForPrintingCompleted): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::drawPagesForPrinting): Remove the callback wrapper.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::print): Remove the printer list hack.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::close): disconnectFromPage() is no longer needed.
(WebKit::WebPage::beginPrinting): Use makeUnique to create the WebPrintOperationGtk.
(WebKit::WebPage::drawPagesForPrinting): Copy the shared buffer to shared memory and send the handle to the UI process.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/gtk/PrinterListGtk.cpp: Removed.
(WebKit::PrinterListGtk::getOrCreate): Deleted.
(WebKit::PrinterListGtk::PrinterListGtk): Deleted.
(WebKit::PrinterListGtk::~PrinterListGtk): Deleted.
(WebKit::PrinterListGtk::addPrinter): Deleted.
(WebKit::PrinterListGtk::findPrinter const): Deleted.
* Source/WebKit/WebProcess/WebPage/gtk/PrinterListGtk.h: Removed.
* Source/WebKit/WebProcess/WebPage/gtk/WebPrintOperationGtk.cpp:
(WebKit::WebPrintOperationGtk::PrintPagesData::PrintPagesData): Make PrintPagesData a private struct in WebPrintOperationGtk
to simplify everyting.
(WebKit::WebPrintOperationGtk::PrintPagesData::incrementPageSequence):
(WebKit::WebPrintOperationGtk::WebPrintOperationGtk):
(WebKit::WebPrintOperationGtk::~WebPrintOperationGtk):
(WebKit::WebPrintOperationGtk::startPrint): Create the cairo surface according to the settings to be written on a stream.
(WebKit::WebPrintOperationGtk::startPage):
(WebKit::WebPrintOperationGtk::endPage):
(WebKit::WebPrintOperationGtk::endPrint):
(WebKit::WebPrintOperationGtk::printPagesIdle):
(WebKit::WebPrintOperationGtk::printPagesIdleDone):
(WebKit::WebPrintOperationGtk::printPagesDone):
(WebKit::WebPrintOperationGtk::printDone):
(WebKit::WebPrintOperationGtk::print):
(WebKit::PrintPagesData::PrintPagesData): Deleted.
(WebKit::PrintPagesData::collatedCopiesLeft): Deleted.
(WebKit::PrintPagesData::uncollatedCopiesLeft): Deleted.
(WebKit::PrintPagesData::copiesLeft): Deleted.
(WebKit::PrintPagesData::incrementPageSequence): Deleted.
(WebKit::WebPrintOperationGtk::create): Deleted.
(WebKit::WebPrintOperationGtk::disconnectFromPage): Deleted.
* Source/WebKit/WebProcess/WebPage/gtk/WebPrintOperationGtk.h:
(WebKit::WebPrintOperationGtk::PrintPagesData::collatedCopiesLeft const):
(WebKit::WebPrintOperationGtk::PrintPagesData::uncollatedCopiesLeft const):
(WebKit::WebPrintOperationGtk::PrintPagesData::copiesLeft const):
(WebKit::WebPrintOperationGtk::printContext const): Deleted.
(WebKit::WebPrintOperationGtk::printSettings const): Deleted.
(WebKit::WebPrintOperationGtk::pageSetup const): Deleted.
(WebKit::WebPrintOperationGtk::printMode const): Deleted.
(WebKit::WebPrintOperationGtk::setNumberOfPagesToPrint): Deleted.
(WebKit::WebPrintOperationGtk::pagesToPrint const): Deleted.
(WebKit::WebPrintOperationGtk::pagePosition const): Deleted.
(WebKit::WebPrintOperationGtk::setPagePosition): Deleted.
(WebKit::WebPrintOperationGtk::pageRanges const): Deleted.
(WebKit::WebPrintOperationGtk::pageRangesCount const): Deleted.
(WebKit::WebPrintOperationGtk::numberUp const): Deleted.
(WebKit::WebPrintOperationGtk::numberUpLayout const): Deleted.
(WebKit::WebPrintOperationGtk::pageSet const): Deleted.
(WebKit::WebPrintOperationGtk::reverse const): Deleted.
(WebKit::WebPrintOperationGtk::copies const): Deleted.
(WebKit::WebPrintOperationGtk::collateCopies const): Deleted.
(WebKit::WebPrintOperationGtk::scale const): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitGtk/TestPrinting.cpp:
(findPrintToFilePrinter):
(testPrintOperationPrint):
(testPrintOperationErrors):
(testPrintOperationCloseAfterPrint):
(testPrintCustomWidget):
(testPrintOperationPrintPrinter): Deleted.

Canonical link: https://commits.webkit.org/253895@main
